Support geohash, geotile and geohex grid types

craigtaverner opened the pull request:
The PR at https://github.com/elastic/elasticsearch/pull/125143 added support for `ST_GEOHASH`, `ST_GEOTILE` and `ST_GEOHEX`. However, since it used `long` as the internal type for the grid id, there was need for many additional functions for converting the `long` to and from `keyword` as well as generating `geo_shape` cell bounds for map display. Each pf these involved many more files (for the functions, their docs and the generated evaluators). With inspiration from PostGIS we decided to take a different direction, and instead use a new internal type for each grid:
* `geohash` for the `ST_GEOHASH` function, created from literal using either `TO_GEOHASH(hash)` or `hash::geohash`
* `geotile` for the `ST_GEOTILE` function, created from literal using either `TO_GEOTILE(tile)` or `tile::geotile`
* `geohex` for the `ST_GEOHEX` function, created from literal using either `TO_GEOHEX(h3)` or `h3::geohex`

This also leads to much stricter type checking as we can no longer use the `long` as a plain `long` in all functions that accept longs, or inadvertently using a geohash in a geotile function. However, the addition of new types involves a lot of boilerplate, especially considering the large number of functions that operate on all types, and need to be informed of the existence of these three new types.

One of the main goals of this work was to also support the concept of a geogrid search. This means "find all documents that intersect a grid cell", similar to the Query DSL version. This is achieved by enabling the use of these three new types in the `ST_INTERSECTS` function, but that will be done in a [followup PR](https://github.com/elastic/elasticsearch/pull/133546).

## Checklist

* [x] Move all new functions to SNAPSHOT
* [x] Move spatial search capabilities into a separate PR (ST_INTERSECTS and related) - https://github.com/elastic/elasticsearch/pull/133546
* [ ] Make PR to move out of SNAPSHOT and link docs
* [ ] Add issues for missing pieces:
  * [ ] Support for shapes
  * [ ] Lucene pushdown
  * [ ] Missing evaluators (literals vs fields, separate evaluators for grid types, etc.)

craigtaverner committed:
Support geohash, geotile and geohex grid types

elasticsearchmachine commented:
Pinging @elastic/es-analytical-engine (Team:Analytics)

craigtaverner committed:
Update docs/changelog/129581.yaml

craigtaverner committed:
Use type-casting syntax to be sure it works

craigtaverner committed:
Merge branch 'esql_geogrid_types' of github.com:craigtaverner/elasticsearch into esql_geogrid_types

craigtaverner committed:
Fix changelog yaml

craigtaverner committed:
New capability due to behaviour change

craigtaverner committed:
Fix failing tests

Added EsqlCapabily to block older BWC
Changed license IT to new API

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into esql_geogrid_types

craigtaverner committed:
Disable broken lucene pushdown

craigtaverner committed:
Merge branch 'main' into esql_geogrid_types

iverase commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/spatial/SpatialRelatesFunction.java:
```diff
@@ -117,6 +133,93 @@ protected boolean geometryRelatesGeometry(GeometryDocValueReader reader, Compone
+                long geoGridId = getGridId(point, gridId, gridType);
+                return gridId == geoGridId;
+            } else {
+                throw new IllegalArgumentException(
```
This should be easy to implement though. It is ok to do it in a follow up

iverase replied:
More over, I think we should do it generic, no point in distinguish a point from another geometry, we have code that that this already.

iverase reviewed (commented):
I like a lot this approach as a grid cell cannot be really represented by a geometry. I am +1 to this approach as far as ESQL folks are good with it.

craigtaverner committed:
Fix failing tests

alex-spies reviewed (changes requested):
~~We need to carefully consider how to deal with the breaking change that this proposes, because we remove some functions altogether.~~ **Update**: Breaking change itself was performed in https://github.com/elastic/elasticsearch/pull/129839, this PR isn't breaking anymore but adds functionality.

Additionally, the new functions added by this aren't snapshot-only, either, and will be made available immediately. It's better to have them bake a little before un-snapshotting them in a dedicated PR.

Finally, this adds multiple data types and wires them immediately - I much prefer we do that in a dedicated PR, this is a big change and can easily introduce bugs.

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into esql_geogrid_types

craigtaverner committed:
Regenerate docs after merging main

craigtaverner committed:
Some fixes after review

craigtaverner committed:
Remove unneeded change

alex-spies commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/spatial-grid.csv-spec:
```diff
@@ -290,27 +376,43 @@ count:long | centroid:geo_point
+// tag::to_geotile_to_long[]
+ROW string = "4/8/5"
+| EVAL geotile = TO_GEOTILE(string)
+| EVAL geotileLong = TO_LONG(geotile)
```
naive question: why do we allow turning geotiles and hashes into longs? Is this conversion canonical and can be used by the user?

I'm wondering if we're unnecessarily exposing an implementation detail here.

alex-spies replied:
Thanks! Makes sense to me.

alex-spies reviewed (commented):
I've started a first pass with the tests.

The ergonomics are certainly much better than before. I have a couple of questions, though.

alex-spies reviewed (commented):
I'm working my way through the PR, but this is a big boy!

Only minor remarks so far, though.

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/action/EsqlCapabilities.java:
```diff
@@ -384,6 +384,11 @@ public enum Cap {
         SPATIAL_GRID(Build.current().isSnapshot()),
 
+        /**
+         * Support geohash, geotile and geohex data types. Done in #129581
```
I was just following the existing pattern, where it seems most people add this comment. I agree that it is redundant, and am not sure why the pattern started, but it does seem to be a kind of convention at the moment.

alex-spies replied:
Oh wow, we're really inconsistent there.

I prefer a link to the original issue than the PR, but I guess both is fine.

ivancea reviewed (commented):
LGTM (Considering my shallow knowledge of geo type conversions).

Added some comments, and looking for Alex comment on breaking changes, which looks relevant here

craigtaverner commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/spatial-grid.csv-spec:
```diff
@@ -179,27 +235,55 @@ FROM airports
+;
+
+count:long | centroid:geo_point
+8          | POINT (6.351574736181647 51.8981519783847)
```
The centroid was validated a few ways, manual calculation of a few cases, and verification that we get the same results from Query DSL (Lucene) and ES|QL in the `SpatialPushDownPointsTestCase`, and that in turn relies on other Lucene and Elasticsearch tests from earlier.

craigtaverner commented on x-pack/plugin/esql/qa/server/single-node/src/javaRestTest/java/org/elasticsearch/xpack/esql/qa/single_node/RestEsqlIT.java:
```diff
@@ -700,7 +703,7 @@ public void testSuggestedCast() throws IOException {
         }
         for (DataType type : shouldBeSupported) {
-            assertTrue(typesAndValues.containsKey(type));
+            assertTrue(type.typeName(), typesAndValues.containsKey(type));
```
We could, but this change seems to give 90% of what we want to see (the type) while remaining concise. I added this because the previous assertion was insufficiently clear.

craigtaverner commented on x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/type/DataType.java:
```diff
@@ -276,6 +276,9 @@ public enum DataType {
     // wild estimate for size, based on some test data (airport_city_boundaries)
     CARTESIAN_SHAPE(builder().esType("cartesian_shape").estimatedSize(200).docValues()),
     GEO_SHAPE(builder().esType("geo_shape").estimatedSize(200).docValues()),
+    GEOHASH(builder().esType("geohash").typeName("GEOHASH").estimatedSize(Long.BYTES)),
+    GEOTILE(builder().esType("geotile").typeName("GEOTILE").estimatedSize(Long.BYTES)),
+    GEOHEX(builder().esType("geohex").typeName("GEOHEX").estimatedSize(Long.BYTES)),
```
Good point. I was going on the fact that we don't really use `esType` to actually mean the type from Elasticsearch (although it might have been intended for that). In particular, the `cartesian_point` and `cartesian_shape` types are not named that in Elasticsearch. I'll remove this hear, and we should separately review the use of this in other cases.

craigtaverner replied:
So I removed, and then brought back the esType. It turns out it is the primary way to describe the column type in the final results. It is used for the `outputType()` method. In fact this appears to be it's primary purpose.

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/spatial/SpatialRelatesFunction.java:
```diff
@@ -117,6 +133,93 @@ protected boolean geometryRelatesGeometry(GeometryDocValueReader reader, Compone
+            }
+        }
+
+        private long getGridId(Point point, @Fixed long gridId, @Fixed DataType gridType) {
```
Yes, these sneaked in with code copying from the evaluator functions. I'll remove them.

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/spatial/StGeohex.java:
```diff
@@ -246,4 +253,23 @@ static void fromFieldDocValuesAndLiteralAndLiteral(
         fromEncodedLong(results, p, encoded, bounds);
     }
+
+    public static BytesRef toBounds(long gridId) {
```
I wanted to keep code specific to the structure of the grids inside the grid function classes. We cannot move this to the utility classes, because this has ES|QL specific code (the fact that we encode the geometries as WKB only in ES|QL), and the utilities are used by both Query DSL and ES|QL (so we'd have a dependency issue). One alternative would be to make three new utilities classes for ES|QL specific behaviour, and have them proxy to the older utilities. But that seems to be adding more complexity, which feels to me like we'd be making things worse.

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/convert/ToGeoShape.java:
```diff
@@ -95,4 +103,9 @@ protected NodeInfo<? extends Expression> info() {
+
+    @ConvertEvaluator(extraName = "FromGeoGrid", warnExceptions = { IllegalArgumentException.class })
+    static BytesRef fromGeoGrid(long in, @Fixed DataType dataType) {
+        return geoGridToShape(in, dataType);
```
I agree, but I've been getting a lot of negative feedback on the generated code, and asked to please find ways to reduce the number of evaluators. This is one idea I had to do that. I have plans for more evaluators anyway going forward, and we can re-visit this question in that PR.

craigtaverner committed:
Code review updates

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into esql_geogrid_types

Commit:
[CI] Auto commit changes from spotless

alex-spies commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/spatial-grid.csv-spec:
```diff
@@ -734,12 +940,12 @@ count:long | centroid:geo_point                            | geohexString:keywor
 | EVAL geohex = ST_GEOHEX(location, 1, TO_GEOSHAPE("BBOX(0.0, 12.0, 60.0, 30.0)"))
-| WHERE ST_GEOHEX_TO_STRING(geohex) == "8108bffffffffff"
-| EVAL geohexString = ST_GEOHEX_TO_STRING(ST_GEOHEX(location, 1))
+| WHERE TO_STRING(geohex) == "8108bffffffffff"
```
Ah, yeah, thank you!

alex-spies commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/spatial-grid.csv-spec:
```diff
@@ -2,27 +2,43 @@
+;
+
+// tag::to_geohash_to_long-result[]
+string:keyword | geohash:geohash | geohashLong:long
+u3bu           | u3bu            | 13686180
```
These test expectations suggest that `geohash` is a type that users will get as return value and can see in the query output when running the query.

However, trying locally with `ROW geohash = (13686180::long)::geohash | EVAL geohashString = TO_STRING(geohash)` I get a `null`-filled column of type `unsupported`.

I assume the grid types aren't supposed to be non-representable? That is, internal types like `DATE_PERIOD` and `TIME_DURATION`? The latter cannot make it into the output of a query.

alex-spies commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/spatial-grid.csv-spec:
Got a 500 from
```
ROW geohash = (13686180)::geohash | EVAL geohashString = TO_STRING(geohash)
```
Seems like our type checking doesn't actually check that I'm feeding a long to `to_geohash`:
```
  java.lang.ClassCastException: class org.elasticsearch.compute.data.ConstantIntVector cannot be cast to class org.elasticsearch.compute.data.LongVector (org.elasticsearch.compute.data
    at org.elasticsearch.xpack.esql.expression.function.scalar.convert.ToStringFromGeoGridEvaluator.evalVector(ToStringFromGeoGridEvaluator.java:44)
    at org.elasticsearch.xpack.esql.expression.function.scalar.convert.AbstractConvertFunction$AbstractEvaluator.eval(AbstractConvertFunction.java:164)
    at org.elasticsearch.xpack.esql.evaluator.mapper.EvaluatorMapper.fold(EvaluatorMapper.java:157)
    at org.elasticsearch.xpack.esql.expression.function.scalar.EsqlScalarFunction.fold(EsqlScalarFunction.java:39)
    at org.elasticsearch.xpack.esql.core.expression.Literal.of(Literal.java:185)
    at org.elasticsearch.xpack.esql.optimizer.rules.RuleUtils.lambda$foldableReferences$1(RuleUtils.java:98)
    at org.elasticsearch.xpack.esql.core.tree.Node.lambda$forEachUp$1(Node.java:100)
    at org.elasticsearch.xpack.esql.core.tree.Node.forEachUp(Node.java:93)
    at org.elasticsearch.xpack.esql.core.tree.Node.forEachUp(Node.java:98)
    at org.elasticsearch.xpack.esql.plan.QueryPlan.lambda$forEachExpressionUp$21(QueryPlan.java:207)
    at org.elasticsearch.xpack.esql.plan.QueryPlan.doForEachExpression(QueryPlan.java:213)
    at org.elasticsearch.xpack.esql.plan.QueryPlan.doForEachExpression(QueryPlan.java:216)
    at org.elasticsearch.xpack.esql.plan.QueryPlan.lambda$forEachExpressionUp$22(QueryPlan.java:207)
    at org.elasticsearch.xpack.esql.core.tree.Node.forEachProperty(Node.java:122)
    at org.elasticsearch.xpack.esql.core.tree.Node.lambda$forEachPropertyUp$3(Node.java:114)
    at org.elasticsearch.xpack.esql.core.tree.Node.forEachUp(Node.java:93)
    at org.elasticsearch.xpack.esql.core.tree.Node.forEachUp(Node.java:91)
    at org.elasticsearch.xpack.esql.core.tree.Node.forEachPropertyUp(Node.java:114)
    at org.elasticsearch.xpack.esql.plan.QueryPlan.forEachExpressionUp(QueryPlan.java:207)
    at org.elasticsearch.xpack.esql.optimizer.rules.RuleUtils.foldableReferences(RuleUtils.java:89)
    at org.elasticsearch.xpack.esql.optimizer.rules.logical.PropagateEvalFoldables.apply(PropagateEvalFoldables.java:28)
    at org.elasticsearch.xpack.esql.optimizer.rules.logical.PropagateEvalFoldables.apply(PropagateEvalFoldables.java:24)
    at org.elasticsearch.xpack.esql.rule.ParameterizedRuleExecutor.lambda$transform$0(ParameterizedRuleExecutor.java:29)
    at org.elasticsearch.xpack.esql.rule.RuleExecutor$Transformation.<init>(RuleExecutor.java:92)
    at org.elasticsearch.xpack.esql.rule.RuleExecutor.executeWithInfo(RuleExecutor.java:171)
    at org.elasticsearch.xpack.esql.rule.RuleExecutor.execute(RuleExecutor.java:140)
    at org.elasticsearch.xpack.esql.optimizer.LogicalPlanOptimizer.optimize(LogicalPlanOptimizer.java:113)
    at org.elasticsearch.xpack.esql.session.EsqlSession.optimizedPlan(EsqlSession.java:1002)
    at org.elasticsearch.xpack.esql.session.EsqlSession$1.lambda$onResponse$0(EsqlSession.java:210)
    at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations$ResponseWrappingActionListener.onResponse(ActionListenerImplementations.java:261) 
    at org.elasticsearch.xpack.esql.planner.premapper.PreMapper.lambda$preMapper$0(PreMapper.java:33)
    at org.elasticsearch.server@9.2.0-SNAPSHOT/org.elasticsearch.action.ActionListenerImplementations$ResponseWrappingActionListener.onResponse(ActionListenerImplementations.java:261) 
    at org.elasticsearch.xpack.esql.expression.function.fulltext.QueryBuilderResolver.resolveQueryBuilders(QueryBuilderResolver.java:50)
    at org.elasticsearch.xpack.esql.planner.premapper.PreMapper.queryRewrite(PreMapper.java:38)
    at org.elasticsearch.xpack.esql.planner.premapper.PreMapper.preMapper(PreMapper.java:31)
    at org.elasticsearch.xpack.esql.session.EsqlSession$1.onResponse(EsqlSession.java:208)
...
```

craigtaverner replied:
Good catch. I thought I was being clever allowing the to_geo* functions to accept integers, but clearly not. I've removed that support now (the alternative would be to make more evaluators that mapped IntBlock to LongBlock).

alex-spies commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/spatial-grid.csv-spec:
Got another interesting validation problem:
```
ROW geohash = to_geotile(to_long(13686180))
```
Gives me my null column of unsupported type.

```
ROW geohash = to_string(to_geotile(to_long(13686180)))
```
gives me a 400 with `Zoom/X/Y combination is not valid: 0/0/13686180`. I think the validation happens at the wrong time; it should already failed at `to_geotile`, but it happily created a tile that doesn't make sense.

Didn't try the same for hex and hash, but all of those require tests that show that our validation fails during the conversion to the grid type, not later when we try to do something with the value.

craigtaverner replied:
The null columns are from the mistake to remove esType. I've put esType back. Should work now.

alex-spies reviewed (commented):
Another round of comments!

My review is not complete yet, but I have a couple of questions:
- **This is a tremendous one-way door** as we add 3 new data types. You also brought up the idea of parameterized types, where hash/hex/tile seems to also fit in: these are of type `long` with metadata "represents a geohex/grid/hash". I wonder: how many more non-parameterized data types are we going to add? Do we need a more general approach to deal with this? I don't know and this change makes sense to me, but I wonder if geohex/tile/grid will stick out like a sore thumb once we go one step further and have to support parameterized geo types anyway.
- Should the new data types be non-representable, like `TIME_DURATION` and `DATE_PERIOD`?
- There's push down logic for functions like `ST_INTERSECTS`; I didn't yet get a chance to check up on it, but does this PR include changes to make sure this still works with the new data types present?
- Are any changes required by downstream clients and Kibana? How big is the change to the query response JSON?

craigtaverner committed:
Fix some errors after moving from esType

craigtaverner committed:
Bring back esType since it is required for outputType

craigtaverner commented on x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/type/DataType.java:
```diff
@@ -276,6 +276,9 @@ public enum DataType {
     GEO_SHAPE(builder().esType("geo_shape").estimatedSize(200).docValues()),
+    GEOHASH(builder().esType("geohash").typeName("GEOHASH").estimatedSize(Long.BYTES)),
+    GEOTILE(builder().esType("geotile").typeName("GEOTILE").estimatedSize(Long.BYTES)),
+    GEOHEX(builder().esType("geohex").typeName("GEOHEX").estimatedSize(Long.BYTES)),
```
OK. That did not work. It turns out esType is required for output type, and without it all results have `unsupported` in integration tests (not CsvTests though). I've put this back to what it was for now, and we should discuss going forward how to deal with types that can really be output, but have no internal mapping.

craigtaverner commented on docs/reference/query-languages/esql/_snippets/functions/types/case.md:
```diff
@@ -20,6 +20,12 @@
 | boolean | geo_point | | geo_point |
 | boolean | geo_shape | geo_shape | geo_shape |
 | boolean | geo_shape | | geo_shape |
+| boolean | geohash | geohash | geohash |
```
Yeah. Autogenerated stuff like this cannot be easily hidden. Will take some hackiness to do so.

craigtaverner commented on docs/reference/query-languages/esql/_snippets/functions/types/st_disjoint.md:
The primary use case here was to do a search on an index with mapping `geo_point` for docs that are within the specified tile literal. We wanted to minimize the number of additional evaluators initially. While I can imagine it being of possible interest to users to ask if a particular geohex intersects a particular geotile, this use case should only be added later, not part of the original work.

alex-spies replied:
Agree, thanks for clarifying!

craigtaverner committed:
Merge branch 'esql_geogrid_types' of github.com:craigtaverner/elasticsearch into esql_geogrid_types

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into esql_geogrid_types

craigtaverner commented on docs/reference/query-languages/esql/_snippets/functions/types/st_disjoint.md:
```diff
@@ -10,6 +10,12 @@
 | cartesian_shape | cartesian_shape | boolean |
 | geo_point | geo_point | boolean |
 | geo_point | geo_shape | boolean |
+| geo_point | geohash | boolean |
```
This was one of the motivations for creating new data types. To reduce the chance people would get confused. An earlier iteration of this actually embedded the fact that a geometry was a grid cell into the geometry WKB, which would allow it to work correctly, but not reduce possible user confusion.

alex-spies replied:
Interesting, thanks for the added context!

craigtaverner committed:
Remove int support from to_geo* functions

The previous support kept the underlying IntBlock, which later would cause exceptions since the type expected a LongBock.

craigtaverner committed:
Merge branch 'main' into esql_geogrid_types

craigtaverner committed:
Merge branch 'main' into esql_geogrid_types

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/convert/ToGeohash.java:
```diff
@@ -16,53 +16,56 @@
         Map.entry(LONG, (source, fieldEval) -> fieldEval),
-        Map.entry(KEYWORD, StGeohashToLongFromStringEvaluator.Factory::new)
+        Map.entry(KEYWORD, ToGeohashFromStringEvaluator.Factory::new),
+        Map.entry(TEXT, ToGeohashFromStringEvaluator.Factory::new)
```
How is converting from text reasonable? Wouldn't text normally be tokenized, and thus could be substantially different from the original source?

alex-spies replied:
Thanks for the context, makes sense!

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/spatial/SpatialDisjoint.java:
The changes to the binary spatial functions really should go into their own PR (or PRs). This is already quite complex to review by itself.

craigtaverner replied:
This has been moved into a followup PR

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/spatial/BinarySpatialFunction.java:
```diff
@@ -150,7 +156,13 @@ protected TypeResolution resolveType() {
         protected Expression.TypeResolution isSpatial(Expression e, TypeResolutions.ParamOrdinal paramOrd) {
```
Making `isSpatial` depend on the function is confusing to me. There is another `isSpatial` method in DataType.java and this has a fixed meaning - now we make the meaning depending on the `supportsGrid` attribute. Maybe this method deserves a better name.

craigtaverner replied:
It seems this function has expanded in meaning in a few steps, so the name has become less appropriate. Already in June 2024, when we added the distance function, it was more complex. I could fix it in this PR, or we treat that separately.

craigtaverner replied:
Renamed to `isCompatibleSpatial`

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/spatial/BinarySpatialFunction.java:
```diff
@@ -182,6 +194,16 @@ protected TypeResolution resolveType(
                 return resolution;
             }
+            // TODO Remove these grid checks once we support geo_shape relation to geoGrid
+            if (resolution.resolved() && DataType.isGeoGrid(spatialExpression.dataType())) {
```
Don't we already check for `resolution.unresolved()` above? Looks redundant.

craigtaverner replied:
OK. I could get rid of one check.

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/join/Join.java:
```diff
@@ -66,6 +69,9 @@ public class Join extends BinaryPlan implements PostAnalysisVerificationAware, S
         GEO_SHAPE,
         CARTESIAN_POINT,
         CARTESIAN_SHAPE,
+        GEOHASH,
```
++

alex-spies reviewed (changes requested):
I'm sorry @craigtaverner, I'd really like to move forward with this PR because I think the approach is correct and the user-facing improvements are great. However, I've spent a good couple hours going over the productive code but I'm afraid this is still just too big for me to adequately review without rubberstamping.

This touches hundreds of files and changes 4k lines, while wiring in the new data types into lots of operators/functions immediately - I think this covers nearly all related to the new data types, except maybe pushdown to Lucene when they are used. I need smaller bites so I can actually think through the whole testing, type resolution etc.

To be able to review this, I'd like to ask you to please:
- Split this up into as small increments as possible while still making sense on their own. I thought of a couple ways to slim down the scope and added them as comments below. I think a minimal start would only include the new data types as well as converters to convert **to** them (and maybe from them).
- Hide all user-facing changes behind SNAPSHOT-only so we can release them into tech preview in a small, separate PR. Also hide the docs changes so we don't spill the docs ahead of time.

Also, I'm still interested in the answers to my questions [above](https://github.com/elastic/elasticsearch/pull/129581#pullrequestreview-3010091262).

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/EsqlFunctionRegistry.java:
```diff
@@ -401,7 +404,10 @@ private static FunctionDefinition[][] functions() {
                 def(StX.class, StX::new, "st_x"),
-                def(StY.class, StY::new, "st_y") },
+                def(StY.class, StY::new, "st_y"),
+                def(StGeohash.class, StGeohash::new, "st_geohash"),
+                def(StGeotile.class, StGeotile::new, "st_geotile"),
+                def(StGeohex.class, StGeohex::new, "st_geohex") },
```
Sorry, my bad, lost track of the new and old functions; however, we still can un-snapshot these in a separate PR.

craigtaverner replied:
They are all in snapshot only now

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into esql_geogrid_types

craigtaverner committed:
Fixed min/max after merging main

craigtaverner committed:
Merge branch 'esql_geogrid_types' of github.com:craigtaverner/elasticsearch into esql_geogrid_types

craigtaverner committed:
Fixed intersects/disjoint after merging main

craigtaverner committed:
Fix after merging main

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into esql_geogrid_types

craigtaverner committed:
Fix tests after merging main

craigtaverner committed:
Fix tests after merging main

craigtaverner committed:
Put grid types back into snapshot and assert that functions only work in snapshot mode

craigtaverner committed:
Move type-conversion functions to snapshot as well

craigtaverner committed:
Fix failing test after moving functions to snapshot

craigtaverner committed:
Commit kibana specs after moving function to snapshot

craigtaverner committed:
Add missing unit tests for new conversion functions

craigtaverner committed:
Add missing docs for to_geo-grid functions

craigtaverner committed:
Merge branch 'main' into esql_geogrid_types

craigtaverner committed:
Simplify type checks

craigtaverner committed:
Redirects to work around build issue in docs-builder

craigtaverner committed:
Redirects to work around build issue in docs-builder

craigtaverner committed:
Using joe to add back trailing whitespace

craigtaverner committed:
Try multi-line redirects

craigtaverner committed:
Try absolute path

craigtaverner committed:
Merge branch 'main' into esql_geogrid_types

craigtaverner committed:
Fixed failing tests after type resolution simplification

The simplification favoured placing the error on the first parameter, but we'd rather place it on the second.

craigtaverner committed:
Return to single-line redirects

craigtaverner commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/spatial-grid.csv-spec:
I was not able to find where we declared them sortable. They should not be sortable.

craigtaverner committed:
Rename function to improve clarity

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into esql_geogrid_types

craigtaverner committed:
Remove support for geogrid in ST_INTERSECTS

craigtaverner committed:
Bring back function supporting grid types in other tests

craigtaverner committed:
Merge branch 'main' into esql_geogrid_types

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into esql_geogrid_types

craigtaverner committed:
Fix evaluators after merging in main

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/EsqlFunctionRegistry.java:
```diff
@@ -403,14 +406,8 @@ private static FunctionDefinition[][] functions() {
-                def(StGeotileToString.class, StGeotileToString::new, "st_geotile_to_string"),
-                def(StGeohex.class, StGeohex::new, "st_geohex"),
-                def(StGeohexToLong.class, StGeohexToLong::new, "st_geohex_to_long"),
-                def(StGeohexToString.class, StGeohexToString::new, "st_geohex_to_string") },
```
moved

craigtaverner commented on docs/reference/query-languages/esql/_snippets/functions/parameters/st_disjoint.md:
```diff
@@ -3,8 +3,8 @@
 
 `geomA`
-:   Expression of type `geo_point`, `cartesian_point`, `geo_shape` or `cartesian_shape`. If `null`, the function returns `null`.
+:   Expression that is either a geometry (`geo_point`, `cartesian_point`, `geo_shape` or `cartesian_shape`) or a geo-grid value (`geohash`, `geotile`, `geohex`). If `null`, the function returns `null`.
```
I've moved everything to SNAPSHOT, moved the intersects feature into a followup PR, and added a test for non-snapshot mode (to the same place I test for different license modes).

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/EsqlFunctionRegistry.java:
```diff
@@ -424,6 +430,9 @@ private static FunctionDefinition[][] functions() {
                 def(ToGeoShape.class, ToGeoShape::new, "to_geoshape"),
+                def(ToGeohash.class, ToGeohash::new, "to_geohash"),
+                def(ToGeotile.class, ToGeotile::new, "to_geotile"),
+                def(ToGeohex.class, ToGeohex::new, "to_geohex"),
```
Done

alex-spies commented on x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/type/DataType.java:
```diff
@@ -567,19 +570,27 @@ public static boolean isSpatialPoint(DataType t) {
+
     public static boolean isSortable(DataType t) {
-        return false == (t == SOURCE || isCounter(t) || isSpatial(t) || t == AGGREGATE_METRIC_DOUBLE);
+        return false == (t == SOURCE || isCounter(t) || isSpatialOrGrid(t) || t == AGGREGATE_METRIC_DOUBLE);
     }
```
Are the grid types supposed to be sortable?

I think they shouldn't be, right? Maybe a small verifier test would help ensure that the validation of `SORT` and `SORT` triggers and flags the grid data types as invalid.

craigtaverner replied:
Correct, they are not sortable. If a user wishes to sort, they can do `TO_STRING` first and sort on that, which makes it clear what they are sorting on. I can look at making a validation test for this.

alex-spies reviewed (approved):
This is lovely. Thanks for splitting this @craigtaverner, much much easier to digest this way.

I found one inconsistency in the behavior of the conversion functions that should be addressed: you cannot convert a geotile type to itself, in contrast to our other data types.

Otherwise I only have minor test suggestions.

Please proceed at your own discretion and feel free to address anything in follow-up PRs if it's easier to get this chonky PR merged, first; I imagine the merge conflicts are not fun to deal with.

craigtaverner committed:
Support self conversions for the three new types

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/aggregate/Count.java:
I investigated and it appears CountDistinct uses `isExact` to ensure that only actual fields (not reference attributes) are ever used. Not sure why this distinction is there, but it does block all types that are not fields (and the grid types are never es fields). So if we ever want to support this in future it is a bigger lift because it means supported non-fields types.

craigtaverner replied:
Sorry, my mistake, I read the code wrong. It does not require fields, but does have explicit support for each type, unlike Count. However, I can see that none of the spatial types are supported. It would be easy to support all spatial types as BytesRef, and grid types as long, but I think we should do that in a separate PR.

craigtaverner committed:
Add csv tests for count(grid)

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/aggregate/Count.java:
Done. Updated the existing tests to use `COUNT(grid)` instead of `COUNT(*)`.

craigtaverner committed:
Added self-conversions to csv-spec tests

craigtaverner committed:
Added CASE example to csv-tests

craigtaverner committed:
Added MV functions to csv-tests

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into esql_geogrid_types

craigtaverner committed:
Support MV_CONTAINS

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/multivalue/MvAppend.java:
Done. Added an MV test for each testing a few MV functions, including the new MV_CONTAINS.

craigtaverner committed:
Add csv-spec tests for COALESCE

craigtaverner committed:
Updated docs after test updates

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/predicate/operator/comparison/In.java:
Oh, dear, this one is a bit of a can of worms, as there seems to be wider support in the operator than in covered by the tests, outside of grid support. Notably datetime. I think a separate issue to resolve all inconsistencies is needed here.
